File: test/domain/insights/cross_metric_insights_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:openvitals/core/time/local_date.dart';
import 'package:openvitals/domain/insights/cross_metric_insights.dart';

final LocalDate _start = LocalDate(2026, 5, 1);

List<CrossMetricValue> _valuesOf(List<double> values) => [
      for (var index = 0; index < values.length; index++)
        CrossMetricValue(date: _start.plusDays(index), value: values[index]),
    ];

void main() {
  test('calculatesPositiveCorrelationForPairedDays', () {
    final primary = _valuesOf([1.0, 2.0, 3.0, 4.0]);
    final secondary = _valuesOf([2.0, 4.0, 6.0, 8.0]);

    final insight = crossMetricInsight(primary, secondary);

    expect(insight, isNotNull);
    expect(insight!.correlation, closeTo(1.0, 0.0001));
    expect(insight.direction, CrossMetricDirection.positive);
    expect(insight.strength, CrossMetricStrength.strong);
    expect(insight.pairedDays, 4);
  });

  test('calculatesNegativeCorrelationForPairedDays', () {
    final primary = _valuesOf([1.0, 2.0, 3.0, 4.0]);
    final secondary = _valuesOf([8.0, 6.0, 4.0, 2.0]);

    final insight = crossMetricInsight(primary, secondary);

    expect(insight, isNotNull);
    expect(insight!.correlation, closeTo(-1.0, 0.0001));
    expect(insight.direction, CrossMetricDirection.negative);
    expect(insight.strength, CrossMetricStrength.strong);
  });

  test('ignoresUnpairedAndEmptyValues', () {
    final primary = _valuesOf([0.0, 2.0, 3.0, 4.0]);
    final secondary = [
      CrossMetricValue(date: _start.plusDays(1), value: 4.0),
      CrossMetricValue(date: _start.plusDays(2), value: 6.0),
      CrossMetricValue(date: _start.plusDays(3), value: 8.0),
      CrossMetricValue(date: _start.plusDays(10), value: 100.0),
    ];

    final insight = crossMetricInsight(primary, secondary);

    expect(insight, isNotNull);
    expect(insight!.pairedDays, 3);
    expect(insight.correlation, closeTo(1.0, 0.0001));
  });

  test('returnsNullWhenThereAreNotEnoughPairs', () {
    final primary = _valuesOf([1.0, 2.0]);
    final secondary = _valuesOf([2.0, 4.0]);

    final insight = crossMetricInsight(primary, secondary);

    expect(insight, isNull);
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/domain/insights/CrossMetricInsightsTest.kt
package tech.mmarca.openvitals.domain.insights

import java.time.LocalDate
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test

class CrossMetricInsightsTest {

    private val start = LocalDate.of(2026, 5, 1)

    @Test
    fun calculatesPositiveCorrelationForPairedDays() {
        val primary = valuesOf(1.0, 2.0, 3.0, 4.0)
        val secondary = valuesOf(2.0, 4.0, 6.0, 8.0)

        val insight = crossMetricInsight(primary, secondary)

        assertNotNull(insight)
        assertEquals(1.0, checkNotNull(insight).correlation, 0.0001)
        assertEquals(CrossMetricDirection.POSITIVE, insight.direction)
        assertEquals(CrossMetricStrength.STRONG, insight.strength)
        assertEquals(4, insight.pairedDays)
    }

    @Test
    fun calculatesNegativeCorrelationForPairedDays() {
        val primary = valuesOf(1.0, 2.0, 3.0, 4.0)
        val secondary = valuesOf(8.0, 6.0, 4.0, 2.0)

        val insight = crossMetricInsight(primary, secondary)

        assertNotNull(insight)
        assertEquals(-1.0, checkNotNull(insight).correlation, 0.0001)
        assertEquals(CrossMetricDirection.NEGATIVE, insight.direction)
        assertEquals(CrossMetricStrength.STRONG, insight.strength)
    }

    @Test
    fun ignoresUnpairedAndEmptyValues() {
        val primary = valuesOf(0.0, 2.0, 3.0, 4.0)
        val secondary = listOf(
            CrossMetricValue(start.plusDays(1), 4.0),
            CrossMetricValue(start.plusDays(2), 6.0),
            CrossMetricValue(start.plusDays(3), 8.0),
            CrossMetricValue(start.plusDays(10), 100.0),
        )

        val insight = crossMetricInsight(primary, secondary)

        assertNotNull(insight)
        assertEquals(3, checkNotNull(insight).pairedDays)
        assertEquals(1.0, insight.correlation, 0.0001)
    }

    @Test
    fun returnsNullWhenThereAreNotEnoughPairs() {
        val primary = valuesOf(1.0, 2.0)
        val secondary = valuesOf(2.0, 4.0)

        val insight = crossMetricInsight(primary, secondary)

        assertNull(insight)
    }

    private fun valuesOf(vararg values: Double): List<CrossMetricValue> =
        values.mapIndexed { index, value ->
            CrossMetricValue(start.plusDays(index.toLong()), value)
        }
}
